### PR TITLE
[BISERVER-14935] import-export.bat/.sh does not work on Pentaho 9.3 or higher if requestParameterAuthenticationEnabled=true

### PR DESCRIPTION
**Pull request by smmribeiro**
Two commits on this PR: the first is the actual fix for this issue's problem; the second is a refactoring of the class to fix some sonar issues and some architectural problems found during the test and development of the first.

@bcostahitachivantara @renato-s @andreramos89

### CODE REVIEW

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 1h 39m 40s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl extensions
```
---
:ok_hand: All tests passed!
 
Tests run: **1693**, Failures: **0**, Skipped: **1** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/platform/job/PR-5417/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java [hunk 1]
```diff
@@ -731,10 +731,18 @@ private void performBackup()
     if ( !authResponse.equals( "true" ) ) {
       System.err.println( Messages.getInstance().getString( "CommandLineProcessor.ERROR_0006_NON_ADMIN_CREDENTIALS" ) );
     }
+
+    // Build the complete URL to use
+    String backupURL = contextURL + API_REPO_FILES_BACKUP;
+
+    // BISERVER-14935: In case "requestParameterAuthenticationEnabled" is set to 'true'
+    backupURL += ( "?userid=" + getUsername() + "&password=" + Encr.encryptPassword(
```
Comment (bcostahitachivantara) [inline extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java:760]: Here you are creating a string on line 736 and then creating a new one. Why not create the string with all data instead of reassigning a new value or use a StringBuilder.

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java [hunk 2]
```diff
@@ -731,10 +731,18 @@ private void performBackup()
...
+      Encr.decryptPasswordOptionallyEncrypted( getPassword() ) ) );
+
     WebResource resource = client.resource( backupURL );
 
     // Response response
-    Builder builder = resource.type( MediaType.MULTIPART_FORM_DATA ).accept( MediaType.TEXT_HTML_TYPE );
+    Builder builder = resource.type( MediaType.APPLICATION_FORM_URLENCODED ).accept( MediaType.TEXT_HTML_TYPE );
```
Comment (bcostahitachivantara) [inline extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java:766]: Why did you change from MULTIPART_FORM_DATA into APPLICATION_FORM_URLENCODED ? Usually MULTIPART_FORM_DATA is more efficient.

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java [hunk 3]
```diff
@@ -346,10 +362,10 @@ private void performREST() throws ParseException, KettleException, URISyntaxExce
         Messages.getInstance().getString( "CommandLineProcessor.INFO_REST_RESPONSE_STATUS", response.getStatus() );
       message += "\n";
 
-      if ( logFile != null && !"".equals( logFile ) ) {
+      if ( logFile != null && !logFile.isEmpty() ) {
```
Comment (bcostahitachivantara) [inline extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java:365]: Why not StringUtils.isEmpty( logFile )?

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java [hunk 4]
```diff
@@ -703,9 +721,9 @@ private void logResponseMessage( String logFile, String path, ClientResponse res
       message.append( Messages.getInstance().getString( "CommandLineProcessor.INFO_REST_RESPONSE_RECEIVED",
         response.getEntity( String.class ) ) );
     }
-    System.out.println( message.toString() );
-    if ( logFile != null && !"".equals( logFile ) ) {
-      writeFile( message.toString(), logFile );
+    System.out.println( message );
+    if ( logFile != null && !logFile.isEmpty() ) {
```
Comment (bcostahitachivantara) [inline extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java:725]: Why not StringUtils.isEmpty( logFile )?

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java [hunk 5]
```diff
@@ -745,17 +766,15 @@ private void performBackup()
...
-      input.close();
+      writeEntityToFile( response, outputFile );
+
       String message = Messages.getInstance().getString( "CommandLineProcessor.INFO_EXPORT_COMPLETED" ).concat( "\n" );
       message += Messages.getInstance().getString( "CommandLineProcessor.INFO_RESPONSE_STATUS", response.getStatus() );
       message += "\n";
-      message += Messages.getInstance().getString( "CommandLineProcessor.INFO_EXPORT_WRITTEN_TO", filename );
-      if ( logFile != null && !"".equals( logFile ) ) {
+      message += Messages.getInstance().getString( "CommandLineProcessor.INFO_EXPORT_WRITTEN_TO", outputFile );
+      if ( logFile != null && !logFile.isEmpty() ) {
```
Comment (bcostahitachivantara) [inline extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java:775]: Why not StringUtils.isEmpty( logFile )?

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java [hunk 6]
```diff
@@ -855,17 +867,14 @@ private void performExport()
...
-      createZipFile( filename, input );
-      input.close();
+      writeEntityToFile( response, outputFile );
       String message = Messages.getInstance().getString( "CommandLineProcessor.INFO_EXPORT_COMPLETED" ).concat( "\n" );
       message += Messages.getInstance().getString( "CommandLineProcessor.INFO_RESPONSE_STATUS", response.getStatus() );
       message += "\n";
-      message += Messages.getInstance().getString( "CommandLineProcessor.INFO_EXPORT_WRITTEN_TO", filename );
-      if ( logFile != null && !"".equals( logFile ) ) {
+      message += Messages.getInstance().getString( "CommandLineProcessor.INFO_EXPORT_WRITTEN_TO", outputFile );
+      if ( logFile != null && !logFile.isEmpty() ) {
```
Comment (bcostahitachivantara) [inline extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java:875]: Why not StringUtils.isEmpty( logFile )?

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java [hunk 7]
```diff
@@ -883,111 +892,21 @@ private boolean isValidExportPath( String filePath, String logFile ) {
...
-        if ( f != null && f.exists() && f.isDirectory() ) {
-          isValid = true;
-        }
+        isValid = f.exists() && f.isDirectory();
       }
     }
 
-    if ( !isValid && logFile != null && !"".equals( logFile ) ) {
-      writeFile( "Invalid file-path:" + filePath, logFile );
+    if ( !isValid && logFile != null && !logFile.isEmpty() ) {
```
Comment (bcostahitachivantara) [inline extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java:903]: Why not StringUtils.isEmpty( logFile )?

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java (hunk 4 above)
Comment (andreramos89) [reply extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java:725]: Deprecate StringUtils.isEmpty(Object) and replace remaining usage (e.g. with ObjectUtils.isEmpty)
https://github.com/spring-projects/spring-framework/issues/25945

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java (hunk 1 above)
Comment (andreramos89) [reply extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java:760]: Same as last one: Deprecate StringUtils.isEmpty(Object) and replace remaining usage (e.g. with ObjectUtils.isEmpty)

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java [hunk 8]
```diff
@@ -1005,14 +924,20 @@ protected synchronized IUnifiedRepository getRepository() throws ParseException,
...
-      throw new ParseException( Messages.getInstance().getErrorString( "CommandLineProcessor.ERROR_0001_MISSING_ARG",
-        option ) );
-    }
-    if ( !emptyOk && StringUtils.isEmpty( value ) ) {
-      throw new ParseException( Messages.getInstance().getErrorString( "CommandLineProcessor.ERROR_0001_MISSING_ARG",
-        option ) );
+    if ( StringUtils.isEmpty( value ) ) {
+      // Is it required?
+      if ( required ) {
+        throw new ParseException( Messages.getInstance().getErrorString( "CommandLineProcessor.ERROR_0001_MISSING_ARG",
```
[the remark this answers is not in the document]
Comment (andreramos89) [reply extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java:930]: I believe "and" would be what Bruno means. It changes the order of the validation

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java (hunk 2 above)
Comment (smmribeiro) [reply extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java:762]: Using MULTIPART_FORM_DATA results in a "no multipart boundary was found".

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java (hunk 3 above)
Comment (smmribeiro) [reply extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java]: Done.

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java (hunk 4 above)
Comment (smmribeiro) [reply extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java]: Done.

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java (hunk 7 above)
Comment (smmribeiro) [reply extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java]: Done.

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java (hunk 6 above)
Comment (smmribeiro) [reply extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java]: Done.

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java (hunk 5 above)
Comment (smmribeiro) [reply extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java]: Done.

File: extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (smmribeiro) [reply extensions/src/main/java/org/pentaho/platform/plugin/services/importexport/CommandLineProcessor.java]: Done.

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 42m 37s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl extensions
```
---
:ok_hand: All tests passed!
 
Tests run: **1693**, Failures: **0**, Skipped: **1** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/platform/job/PR-5417/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Review by bcostahitachivantara — APPROVED** [review]
Looks good!
